epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyVisibilityItem.kt

```diff
@@ -0,0 +1,208 @@
+package com.airbnb.epoxy
+
+import android.graphics.Rect
+import android.view.View
+import android.view.ViewGroup
+import androidx.annotation.IntRange
+import androidx.annotation.Px
+import androidx.annotation.VisibleForTesting
+import androidx.recyclerview.widget.RecyclerView
+
+/**
+ * This class represent an item in a [android.view.ViewGroup] and it is
+ * being reused with multiple model via the update method. There is 1:1 relationship between an
+ * EpoxyVisibilityItem and a child within the [android.view.ViewGroup].
+ *
+ * It contains the logic to compute the visibility state of an item. It will also invoke the
+ * visibility callbacks on [com.airbnb.epoxy.EpoxyViewHolder]
+ *
+ * This class should remain non-public and is intended to be used by [EpoxyVisibilityTracker]
+ * only.
+ */
+@VisibleForTesting(otherwise = VisibleForTesting.PACKAGE_PRIVATE)
+class EpoxyVisibilityItem {
+    private val localVisibleRect = Rect()
+    var adapterPosition = RecyclerView.NO_POSITION
+        private set
+
+    @Px
+    private var height = 0
+
+    @Px
+    private var width = 0
+
+    @Px
+    private var visibleHeight = 0
+
+    @Px
+    private var visibleWidth = 0
+
+    @Px
+    private var viewportHeight = 0
+
+    @Px
+    private var viewportWidth = 0
+    private var partiallyVisible = false
+    private var fullyVisible = false
+    private var visible = false
+    private var focusedVisible = false
+    private var viewVisibility = View.GONE
+
+    /** Store last value for de-duping  */
+    private var lastVisibleHeightNotified = NOT_NOTIFIED
+    private var lastVisibleWidthNotified = NOT_NOTIFIED
+    private var lastVisibilityNotified = NOT_NOTIFIED
+
+    constructor() {}
+    constructor(adapterPosition: Int) {
+        reset(adapterPosition)
+    }
```

Comment:
We could also move the empty constructor to the class definition but no strong preference either way here. That'd looks something like 
```
class EpoxyVisibilityItem() {
    ...
    constructor(adapterPosition: Int) : this() {
        reset(adapterPosition)
    }
```

Reply:
Oh, that was generated by the IDE

Reply:
Heh, yeah I figured :-) Auto conversion can do some odd stuff sometimes.